### DISCUSSION TITLE
`.parse()` Typescript linter accepts anything (string, number...)

### DISCUSSION THREAD

**Discussion by sneko** [colinhacks/zod#1894] [Q&A]
Hi,

I didn't find someone having my issue so I'm probably doing something wrong but... I can't figure the solution out.

Let's say I have:
```ts
const TestSchema = z.object({
  email: z.string(),
});

TestSchema.parse('hello');
```

The last line will not throw an error with the Typescript whereas when hovering the parse function I can see:
```ts
(method) ZodType<{ email: string; }, ZodObjectDef<{ email: ZodString; }, "strip", ZodTypeAny>, { email: string; }>.parse(data: unknown, params?: Partial<z.ParseParams> | undefined): {
    email: string;
}
```

**Note that I use `strict: true` in my `tsconfig.json` (and at runtime, `zod` throws an error obviously).**

A "workaround" would be to always do this:
```ts
const TestSchema = z.object({
  email: z.string(),
});
type TestSchemaType = z.infer<typeof TestSchema>;

TestSchema.parse(<TestSchemaType>'hello');
```
... but it's pretty verbose and annoying.

Any idea what's the issue to get the linter working normally?

Thank you,

EDIT: it seems a solution would be to use
```ts
const typedParse = <Schema extends z.ZodSchema> (
  schema: Schema, data: z.infer<Schema>
): z.infer<Schema> => schema.parse( data )

typedParse(ATest, 'hello');
```
(ref: https://github.com/colinhacks/zod/issues/1748#issuecomment-1364572126)

... but I don't get why this is not used by default?

**Comment by JacobWeisenburger** [comment]
> Any idea what's the issue to get the linter working normally?

What do you mean by 'normally'?

---

> but I don't get why this is not used by default?

https://github.com/colinhacks/zod/issues/1748#issuecomment-1364573095

---

> ```ts
> const typedParse = <Schema extends z.ZodSchema> (
>   schema: Schema, data: z.infer<Schema>
> ): z.infer<Schema> => schema.parse( data )
> 
> typedParse(ATest, 'hello');
> ```

Another way you can solve this problem:

```ts
type TypedParseSchema<Schema extends z.ZodSchema> = Omit<Schema, 'parse'> & {
    parse ( data: z.infer<Schema> ): z.infer<Schema>
}
const useTypedParse = <Schema extends z.ZodSchema>
    ( schema: Schema ): TypedParseSchema<Schema> => schema

const schemaWithTypedParse = useTypedParse( z.object( {
    email: z.string(),
} ) )

schemaWithTypedParse.parse( 'hello' )
//                          ^^^^^^^
// Argument of type 'string' is not assignable to parameter of type '{ email: string; }'.
```

Would this work for you?

**Comment by colinhacks** [comment]
This is now achievable with the new `.decode()` methods in Zod 4.1, which were added in conjunction with the Codec API: https://colinhacks.com/essays/introducing-zod-codecs
